src/backend_model.cc

```diff
@@ -88,67 +83,115 @@ BackendModel::BackendModel(
   {
     common::TritonJson::Value optimization;
     if (model_config_.Find("optimization", &optimization)) {
       common::TritonJson::Value pinned_memory;
       if (optimization.Find("input_pinned_memory", &pinned_memory)) {
-        THROW_IF_BACKEND_MODEL_ERROR(
+        RETURN_IF_ERROR(
             pinned_memory.MemberAsBool("enable", &enable_pinned_input_));
       }
       if (optimization.Find("output_pinned_memory", &pinned_memory)) {
-        THROW_IF_BACKEND_MODEL_ERROR(
+        RETURN_IF_ERROR(
             pinned_memory.MemberAsBool("enable", &enable_pinned_output_));
       }
     }
   }
 
-  THROW_IF_BACKEND_MODEL_ERROR(
+  RETURN_IF_ERROR(
       BatchInput::ParseFromModelConfig(model_config_, &batch_inputs_));
-  THROW_IF_BACKEND_MODEL_ERROR(
+  RETURN_IF_ERROR(
       BatchOutput::ParseFromModelConfig(model_config_, &batch_outputs_));
   for (const auto& batch_output : batch_outputs_) {
     for (const auto& name : batch_output.TargetNames()) {
       batch_output_map_.emplace(name, &batch_output);
     }
   }
   triton::common::TritonJson::Value config_inputs;
-  THROW_IF_BACKEND_MODEL_ERROR(
-      model_config_.MemberAsArray("input", &config_inputs));
+  RETURN_IF_ERROR(model_config_.MemberAsArray("input", &config_inputs));
   for (size_t i = 0; i < config_inputs.ArraySize(); i++) {
     triton::common::TritonJson::Value io;
-    THROW_IF_BACKEND_MODEL_ERROR(config_inputs.IndexAsObject(i, &io));
+    RETURN_IF_ERROR(config_inputs.IndexAsObject(i, &io));
     std::string io_name;
-    THROW_IF_BACKEND_MODEL_ERROR(io.MemberAsString("name", &io_name));
+    RETURN_IF_ERROR(io.MemberAsString("name", &io_name));
     triton::common::TritonJson::Value input_property_json;
     bool allow_ragged_batch = false;
     if (io.Find("allow_ragged_batch", &input_property_json)) {
-      THROW_IF_BACKEND_MODEL_ERROR(
-          input_property_json.AsBool(&allow_ragged_batch));
+      RETURN_IF_ERROR(input_property_json.AsBool(&allow_ragged_batch));
     }
     if (allow_ragged_batch) {
       ragged_inputs_.emplace(io_name);
     }
     bool optional = false;
     if (io.Find("optional", &input_property_json)) {
-      THROW_IF_BACKEND_MODEL_ERROR(input_property_json.AsBool(&optional));
+      RETURN_IF_ERROR(input_property_json.AsBool(&optional));
     }
     if (optional) {
-      if (allow_optional) {
+      if (allow_optional_) {
         optional_inputs_.emplace(io_name);
       } else {
-        THROW_IF_BACKEND_MODEL_ERROR(TRITONSERVER_ErrorNew(
+        RETURN_IF_ERROR(TRITONSERVER_ErrorNew(
             TRITONSERVER_ERROR_INVALID_ARG,
             (std::string("'optional' is set to true for input '") + io_name +
              "' while the backend model doesn't support optional input")
                 .c_str()));
       }
     }
   }
+
+  return nullptr;
 }
 
+TRITONSERVER_Error*
+BackendModel::SetModelConfig()
+{
+  triton::common::TritonJson::WriteBuffer json_buffer;
+  ModelConfig().Write(&json_buffer);
```

Comment:
tritonjson::Write returns TRITONSERVER_Error* for invalid json, so we should check it here with RETURN_IF_ERROR